cpp/src/cider/exec/nextgen/function/CiderStringFunctions.cpp

```diff
@@ -199,7 +248,69 @@ extern "C" ALWAYS_INLINE int64_t cider_rconcat(char* string_heap_ptr,
   return pack_string_t(s);
 }
 
-extern "C" ALWAYS_INLINE int8_t* get_buffer_with_realloc_on_demand(
+extern "C" ALWAYS_INLINE int32_t cider_rconcat_len(int lhs_len, int rhs_len) {
+  return lhs_len + rhs_len;
+}
+
+extern "C" ALWAYS_INLINE void cider_rconcat_ptr(char* buffer_ptr,
+                                                const char* lhs,
+                                                int lhs_len,
+                                                const char* rhs,
+                                                int rhs_len) {
+  memcpy(buffer_ptr, rhs, rhs_len);
+  memcpy(buffer_ptr + rhs_len, lhs, lhs_len);
+}
+
+extern "C" RUNTIME_EXPORT int8_t* get_buffer_without_realloc(const int8_t* input_desc_ptr,
+                                                             const int32_t index) {
+  const ArrowArray* arrow_array = reinterpret_cast<const ArrowArray*>(input_desc_ptr);
+  CiderArrowArrayBufferHolder* holder =
+      reinterpret_cast<CiderArrowArrayBufferHolder*>(arrow_array->private_data);
+
+  return holder->getBufferAs<int8_t>(index);
+}
+
+extern "C" ALWAYS_INLINE void copy_string_buffer(const int8_t* input_desc_ptr,
+                                                 int8_t* buffer,
+                                                 int64_t total_row) {
+  const int64_t* offset_buffer = reinterpret_cast<const int64_t*>(
+      reinterpret_cast<const ArrowArray*>(input_desc_ptr)->buffers[1]);
+  int32_t* actual_writable_offset_buffer =
+      reinterpret_cast<int32_t*>(const_cast<int64_t*>(offset_buffer));
+  int32_t cur_offset = 0;
+  for (int i = 0; i < total_row; i++) {
+    int32_t cur_len = (offset_buffer[i] >> 48);
```

Comment:
encode length and offset into one int64_t? maybe you should define some util macro explicitly.
and maybe 32 for length and offset be better?

Reply:
it's string pointer and length.

Reply:
If it's an actual pointer， then encoding the pointer into 48bit is risky.

Reply:
The [x86-64 architecture](https://en.wikipedia.org/wiki/X86-64) (as of 2016) allows 48 bits for virtual memory and, for any given processor, up to 52 bits for physical memory.[[27]](https://en.wikipedia.org/wiki/64-bit_computing#cite_note-28)[[28]](https://en.wikipedia.org/wiki/64-bit_computing#cite_note-29). I think there is no risk for recent CPUs.